protocol/mqtt_paho/v2/protocol.go

```diff
@@ -71,14 +72,20 @@ func (t *Protocol) Send(ctx context.Context, m binding.Message, transformers ...
 	var err error
 	defer m.Finish(err)
 
-	topic := cecontext.TopicFrom(ctx)
-	if topic == "" {
-		topic = t.senderTopic
+	topic := t.senderTopic
+	retain := t.retained
+	qos := t.qos
+
+	protocolCtx := ProtocolContextFrom(ctx)
+	if !reflect.DeepEqual(protocolCtx, ProtocolContext{}) {
```

Comment:
Just want to make sure that you meant for this to be an all-or-nothing step. Meaning, the context can't set just one of he three without clearing the other two values. Was that intended ?

Reply:
Agree, let's return nil if the value cannot be found (follows Go ctx conventions IMHO) and avoid reflect here by simply using a `if protocolCtx != nil` check

Reply:
Also, can we get some coverage for these changes?

Reply:
@duglin Yes. 
If we just set the `Topic` in `ProtocolContext`, then `QoS` and `Retained` will use the default `0 `and `false`, it seems that we can't distinguish whether they are set by the user or used by the default. 
So I deal with it like this here. If you use ProtocolContext, I think you have set these three values. do you have a better suggestion?

Reply:
@embano1 I used the `ProtocolContext` in the integration test to verify that it works as expected.

Reply:
For topic/string we can look for a non-empty string to know if they set it in the context. But for `qos` and `retained` since the default values for those types are also valid mqtt values I guess we're stuck.

One option I can think of would be to split them out from the struct and create separate contexts for each - but that seems like it might be overkill.

Another would is to define them as pointers (e.g. `*bool` so if it's `nil` we know it's not set. That'll work but it is more typing for the user and only something we should consider if we think a popular usecase will be for people to only set a subset of the 3 flags. I don't have enough experience with MQTT to know if that's the case or not.

Reply:
Thanks @duglin! I think it's a better idea to define them as pointers. This practice also seems to be common in [paho mqtt](https://github.com/eclipse/paho.golang/blob/86185b4d2a6c372afe12535368e2dcf805278911/paho/cp_publish.go#L27-L30). We can also provide [a constructor](https://github.com/cloudevents/sdk-go/pull/924/files#diff-c9bfbe8b1a4e23643b340cf417da6912def99e374fab7bb4d95486f87da8fd8dR18-R25) to reduce user typing for these pointer variables.